include/inexor/vulkan-renderer/io/byte_stream.hpp

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include "inexor/vulkan-renderer/world/cube.hpp"
+#include "inexor/vulkan-renderer/world/indentation.hpp"
+
+#include <array>
+#include <filesystem>
+#include <utility>
+#include <vector>
+
+namespace inexor::vulkan_renderer::io {
+class ByteStream {
+protected:
+    std::vector<std::uint8_t> m_buffer;
+
+    /// Read from file.
+    [[nodiscard]] static std::vector<std::uint8_t> read_file(const std::filesystem::path &path);
+
+public:
+    ByteStream() = default;
+    explicit ByteStream(std::vector<std::uint8_t> buffer);
+    /// Read from file.
+    explicit ByteStream(const std::filesystem::path &path);
+
+    [[nodiscard]] std::size_t size() const;
+    [[nodiscard]] const std::vector<std::uint8_t> &buffer() const;
+};
+
+class ByteStreamReader {
+private:
+    const ByteStream &m_stream;
+    /// Stream iterator.
+    std::vector<std::uint8_t>::const_iterator m_iter;
+
+    void check_end(const std::size_t size) const;
+
+public:
+    explicit ByteStreamReader(const ByteStream &stream);
+
+    [[nodiscard]] std::size_t remaining() const;
+    void skip(const std::size_t size);
+
+    /// Generic read method.
+    template <class T, class... Args>
+    [[nodiscard]] T read(const Args &... args);
```

Comment:
```suggestion
    [[nodiscard]] T read(const Args &...);
```
This silences a few warnings in some of the specializations too (todo with non-matching param name).